### PR TITLE
[GEOT-6360] add preffered image format support in WMSLayer and

### PR DESCRIPTION
**Pull request by imranrajjad**
WMSCoverageReader

added constructors that take preffered WMS image format for remote wms
service

https://osgeo-org.atlassian.net/browse/GEOT-6360

### CODE REVIEW

**Comment by nmco** [comment]
The QA failure is valid, related with code formatting:
```
[INFO] Web Map Service Module ............................. FAILURE [02:02 min]
```

Please fix this @imranrajjad :smile:

**Comment by imranrajjad** [comment]
@nmco and @aaime, please review and merge

**Comment by nmco** [comment]
The PR description looks like a few notes that we need to figure out how to merge together, a more constructive description would be appreciated. IMHO the issue should describe the problem, and if the issue doesn't contain it already, the PR description shoudl quickly summarize the intended fix or link to any relevant ML discussion. This would be really helpful :smiley:.

File: modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSCoverageReader.java [hunk 1]
```diff
@@ -133,6 +133,32 @@ public WMSCoverageReader(WebMapServer wms, Layer layer, String style) {
...
     }
 
+    public WMSCoverageReader(WebMapServer wms, Layer layer, String style, String preferredFormat) {
+        this.wms = wms;
+
+        // init the reader
+        addLayer(layer, style);
+
+        // best guess at the format with a preference for PNG (since it's normally transparent)
+        List<String> formats = wms.getCapabilities().getRequest().getGetMap().getFormats();
```
Comment (nmco) [inline modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSCoverageReader.java]: The comment looks out of context. You are basically getting the supported formats form the remote and then cheeking that the pass by format is supported.

File: modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSLayer.java [hunk 2]
```diff
@@ -88,6 +88,18 @@ public WMSLayer(WebMapServer wms, Layer layer, String style) {
...
     }
 
+    /**
+     * Builds a new WMS layer
+     *
+     * @param wms
+     * @param layer
+     * @param style
+     * @param format
+     */
```
Comment (nmco) [inline modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSLayer.java]: Yes that's what a constructor usually do, this Javadoc doesn't add any valuable information.

File: modules/extension/wms/src/test/resources/org/geotools/ows/wms/map/caps130.xml [hunk 3]
```diff
@@ -23,6 +23,7 @@
       </GetCapabilities>
       <GetMap>
         <Format>image/png</Format>
```
Comment (nmco) [inline modules/extension/wms/src/test/resources/org/geotools/ows/wms/map/caps130.xml]: What will happen if `ìmage/png` is not supported by the cascaded server?

File: modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSCoverageReader.java [hunk 4]
```diff
@@ -133,6 +133,32 @@ public WMSCoverageReader(WebMapServer wms, Layer layer, String style) {
...
+        this.format = preferredFormat;
+        // verify if prefferedFormat is supported
+        if (!formats.contains(preferredFormat)) {
+            // if prefereded format is not supported default to first available
+            LOGGER.warning(preferredFormat + " format is not supported, falling back to defaults");
+            for (String format : formats) {
```
Comment (nmco) [inline modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSCoverageReader.java]: This code is an exact copy paste of the code just 20 lines above, can we avoid this :fearful: ?

**Comment by nmco** [comment]
A more general comment, I can definitely see this as an improvement, but I can also foreseen future improvements, for example, like passing the CRS the caller is interested in. If we follow the current path of adding separate elements to the constructor ... well we may end up with a significant number of constructors.

Why no creating a parameters aggregator class? MapSpec? RequestParams? Can you have a look at GeoServer WMS implementation for ideas?

File: modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSCoverageReader.java (hunk 4 above)
Comment (imranrajjad) [reply modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSCoverageReader.java]: sure I will move this code into a utility function

File: modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSLayer.java (hunk 2 above)
Comment (imranrajjad) [reply modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSLayer.java]: will remove it

File: modules/extension/wms/src/test/resources/org/geotools/ows/wms/map/caps130.xml (hunk 3 above)
Comment (imranrajjad) [reply modules/extension/wms/src/test/resources/org/geotools/ows/wms/map/caps130.xml]: the current implementation assumes that the cascaded server will provide image/png as a bare mimimum. With the currently available constructor is if `image/png` is not supported then the WMS requests will probably go incomplete, shall I throw an Exception for this?

Comment (imranrajjad) [reply modules/extension/wms/src/test/resources/org/geotools/ows/wms/map/caps130.xml]: The previous implementation expects `ìmage/png` as bare minimum to be available, thats why it always defaults to `image/png`

File: modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSCoverageReader.java (hunk 1 above)
Comment (imranrajjad) [reply modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSCoverageReader.java]: fixed it

**Comment by imranrajjad** [comment]
@nmco comments addressed

**Comment by imranrajjad** [comment]
@nmco can this be merged?

**Comment by nmco** [comment]
Looks good, thx.

**Pull request merged**